src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc

```diff
@@ -568,18 +602,35 @@ void BootstrapRequest<I>::image_sync() {
                                                m_local_mirror_uuid, m_journaler,
                                                m_client_meta, ctx,
 					       m_progress_ctx);
-  request->start();
+  {
+    Mutex::Locker locker(m_lock);
+    request->get();
+    m_image_sync_request = request;
+  }
+
+  request->send();
 }
 
 template <typename I>
 void BootstrapRequest<I>::handle_image_sync(int r) {
   dout(20) << ": r=" << r << dendl;
 
+  {
+    Mutex::Locker locker(m_lock);
+    m_image_sync_request->put();
+    m_image_sync_request = nullptr;
+  }
+
   if (r < 0) {
     derr << ": failed to sync remote image: " << cpp_strerror(r) << dendl;
     m_ret_val = r;
   }
 
+  if (m_canceled) {
```

Comment:
Perhaps move this above the preceding ```if``` statement since ```r``` most likely equals ```-ECANCELED``` and we don't need an error message.